consensus/src/quorum_store/batch_requester.rs

```diff
@@ -102,71 +100,46 @@ impl<T: QuorumStoreSender> BatchRequester<T> {
-            .await;
-    }
-
-    pub(crate) async fn add_request(
-        &mut self,
+    pub(crate) fn request_batch(
+        &self,
         digest: HashValue,
         signers: Vec<PeerId>,
-        ret_tx: oneshot::Sender<Result<Vec<SignedTransaction>, aptos_executor_types::Error>>,
+        ret_tx: oneshot::Sender<Result<Vec<SignedTransaction>, Error>>,
     ) {
         let mut request_state = BatchRequesterState::new(signers, ret_tx);
-        let request_peers = request_state
-            .next_request_peers(self.request_num_peers)
-            .unwrap(); // note: this is the first try
-
-        debug!("QS: requesting from {:?}", request_peers);
+        let network_sender = self.network_sender.clone();
+        let request_num_peers = self.request_num_peers;
+        let my_peer_id = self.my_peer_id;
+        let epoch = self.epoch;
+        let timeout = Duration::from_millis(self.request_timeout_ms as u64);
 
-        self.digest_to_state.insert(digest, request_state);
-        self.send_requests(digest, request_peers).await;
-        self.timeouts.add_digest(digest, self.request_timeout_ms);
-    }
-
-    pub(crate) async fn handle_timeouts(&mut self) {
-        for digest in self.timeouts.expire() {
-            debug!("QS: timed out batch request, digest = {}", digest);
-            if let Some(state) = self.digest_to_state.get_mut(&digest) {
-                if let Some(request_peers) = state.next_request_peers(self.request_num_peers) {
-                    // Quorum Store measurements
-                    counters::SENT_BATCH_REQUEST_RETRY_COUNT.inc();
-                    self.send_requests(digest, request_peers).await;
-                    self.timeouts.add_digest(digest, self.request_timeout_ms);
-                } else {
-                    let state = self.digest_to_state.remove(&digest).unwrap();
-                    state.serve_request(digest, None);
+        tokio::spawn(async move {
+            while let Some(request_peers) = request_state.next_request_peers(request_num_peers) {
+                let mut futures = FuturesUnordered::new();
+                debug!("QS: requesting from {:?}", request_peers);
+                let request = BatchRequest::new(my_peer_id, epoch, digest);
+                for peer in request_peers {
+                    counters::SENT_BATCH_REQUEST_COUNT.inc();
+                    futures.push(network_sender.request_batch(request.clone(), peer, timeout));
                 }
+                while let Some(response) = futures.next().await {
+                    if let Ok(batch) = response {
+                        counters::RECEIVED_BATCH_COUNT.inc();
+                        if batch.verify().is_ok() {
+                            let digest = batch.digest();
+                            let payload = batch.into_payload();
+                            request_state.serve_request(digest, Some(payload));
+                            return;
```

Comment:
We drop the rest of the futures, so probably we drop an internal oneshot channel inside the RPC implementation. We do not panic in this case, right? (I could not find it in the networking code).

Reply:
it's here https://github.com/aptos-labs/aptos-core/blob/f005fa18b1095aad2ba4f128e5da88c89e4d164b/network/src/protocols/rpc/mod.rs#L400